### ISSUE TITLE
Tuning loss weighting for single class settings

### ISSUE THREAD

**Issue by nemtiax** [google/automl#812]
I have a dataset with a single class, and I'm having some trouble with false positive detections.  My suspicion is that this might be due to some hyperparameters being tuned for the full 80-class setting of COCO, that maybe don't transfer too well to a single class.  Perhaps with only two classes (the object class and the 'background' dummy class), gradients from class loss tend to be smaller, and so the network focuses more on the box loss?  And so I should try decreasing the box_loss_weight parameter?  Or maybe I need to look at the focal loss parameters?  I don't have a great intuition for what the implications of changing those would be, though.

Does anyone have experience working in the single class setting and adjusting these parameters?  In the worst case I could do a grid search, but that quickly becomes too expensive for me from a compute perspective.

**Comment by Steve-2040** [comment]
I had some improvement of false positive detection in the past by adding a second class that I don’t actually use. I think it may work by effectively adding more background that don’t contain the class that have the false positive detections. I know this isn’t a direct solution but it helped me in a similar situation.

**Comment by kartik4949** [comment]
@mingxingtan I also have similar question. 
any thoughts?

**Comment by nemtiax** [comment]
> I had some improvement of false positive detection in the past by adding a second class that I don’t actually use. I think it may work by effectively adding more background that don’t contain the class that have the false positive detections. I know this isn’t a direct solution but it helped me in a similar situation. Did you try to add background training images that have no objects (e.g. no bbox)?

I have quite a few training images with no objects, although it's possible that part of my issue is that the 'background' in these images is relatively static.  I'm doing person detection, and my data only has dozens of rooms across thousands of frames, so it may be that I'm not getting enough variety of negative examples.  My test set has different rooms, so it's possible the network just doesn't know what to make of random objects in those unseen test rooms.

On the other hand, I'm seeing a (subjectively) surprising amount of false positives even on the training data, which is why I'm suspicious that I may need to rebalance training to focus more on correctly differentiating background from person.  The box coordinates on both training and test true positives are impeccable, so I'm wondering if perhaps training was overly focused on box loss.  It's also possible, of course, that box regression is just fundamentally pretty easy in my data, and that's why the boxes are so good.

**Comment by Steve-2040** [comment]
My knowledge on the internal workings of the networks is limited but I had similar issues with the object_detection API using only one class. This improved when using more classes. EfficientDet performance for me, is better regarding false positives but I have not tested it with only one class. It will be good if you get comments from EfficientDet users using only one class regarding performance.

**Comment by nemtiax** [comment]
One hypothesis I'm looking at is that perhaps because my training data consists of a limited number of mostly-static backgrounds, the network is spending too much time getting too good at recognizing those scenes as background.  It seems like turning up the focal loss gamma parameter might encourage it to instead focus more on the harder instances of the person class, and less on super easy backgrounds.

**Comment by Steve-2040** [comment]
That does sound like a good explanation. I am thinking using more class may give the network a deeper understanding of the visual world around it. It may therefore effectively train easier, when using more classes, to a higher level of abstraction. Can you not split the person images into e.g. head, body and complete person 3 class? Then in post processing combine that into a meaningful result. There may be a possibility that these networks rely on more classes to be effective?

**Comment by Steve-2040** [comment]
I am thinking if the theory of more classes to be effective is true then there should be a big difference between training from scratch and only fine-tuning a single class because the fine-tuned model has already a high level of abstraction from the original training. I always train from scratch so have no comparison data between the two methods regarding false positives for a single class.
